src/twisted/web/_flatten.py

```diff
@@ -381,14 +385,40 @@ async def _flattenTree(
 
     @return: A C{Deferred}-returning coroutine that resolves to C{None}.
     """
+    buf = []
+    bufSize = 0
+
+    # Accumulate some bytes up to the buffer size so that we don't annoy the
+    # upstream writer with a million tiny string.
+    def bufferedWrite(bs: bytes) -> None:
+        nonlocal bufSize
+        buf.append(bs)
+        bufSize += len(bs)
+        if bufSize >= BUFFER_SIZE:
+            flushBuffer()
+
+    # Deliver the buffered content to the upstream writer as a single string.
+    # This is how a "big enough" buffer gets delivered, how a buffer of any
+    # size is delivered before execution is suspended to wait for an
+    # asynchronous value, and how anything left in the buffer when we're
+    # finished is delivered.
+    def flushBuffer() -> None:
+        nonlocal bufSize
+        if bufSize > 0:
+            write(b"".join(buf))
+            del buf[:]
+            bufSize = 0
+
     stack: List[Generator] = [
-        _flattenElement(request, root, write, [], None, escapeForContent)
+        _flattenElement(request, root, bufferedWrite, [], None, escapeForContent)
     ]
+
     while stack:
         try:
             frame = stack[-1].gi_frame
             element = next(stack[-1])
             if isinstance(element, Deferred):
+                flushBuffer()
```

Comment:
I think that this is good enough.

```suggestion
                # For now, keep it simple and just flush on any deferred.
                # As a further optimization, left for future changes,
                # If deferred is already called and deferred.result
                # is not a defered, we can keep adding to the buffer,
                # without flushing it. 
                flushBuffer()
```
Maybe just add a comment for someone looking into further optimization of this part.

Reply:
Eh, I don't want to add anything about the `result` attribute.  The only people I want looking at that don't need any hints from me.